Check intent attached to order before processing payment

htdat opened the pull request:
Fixes #5187

## Changes proposed in this Pull Request

- Check the status of attached intent during the payment processing.
- Mark the order successful if the intent is one of successful statuses, and redirect customers to the order-received of this order.

## Testing instructions

### Prep 

- Run `npm run watch` to build JS files.
- Feel free to test in multiple flows and different type cards: 
    - UPE and non-UPE flows.
    - classic and block checkout pages.
    - basic and 3DS cards. 
- Add this snippet to your site https://gist.github.com/htdat/6a52eec1be77b8442ac05401908a3627. You can add it with plugin `Code Snippets` or directly to the bottom of file `woocommerce-payments.php` in the root directory of this project.
- For basic (non-3DS) card in the non-UPE flow, we need to inject this diff to `woocommerce-payments.php` https://gist.github.com/htdat/09eef1ed0cbeec50a5354ca8da38d623 

### Start 

- Add some products, visit to the checkout page.
- Try to make an payment in the checkout page.
- See an error due to our snippets (trying to replicate potential issues that real customers can get).
- See this order (usually the newest order) in WP Admin > Order, confirm that it has not been paid.
- Try to make another payment again.
- Get redirected to `order-received` page of this order with this parameter `wcpay_previous_successful_intent=yes`, e.g.: http://localhost:8082/checkout/order-received/zyz/?key=wc_order_1b9oTbcXtWH0l&wcpay_previous_successful_intent=yes

Screenshot:
![Markup on 2023-01-06 at 17:35:02](https://user-images.githubusercontent.com/10045087/210990591-76f33dc7-711a-4527-a3b0-c7d19f4b2ba4.png)


<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)


**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : [link](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-5.3.0#check-intent-attached-to-order-before-processing-payment-pr-5346)
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

htdat committed:
Add check_intent_attached_to_order_succeeded to both the main gateway and UPE classes.

htdat committed:
Try to reuse the existing intent id.

- Update the main gateway to try reusing the attached intent_id
- Change from create_and_confirm_intention
to create_or_update_intention_with_confirmation: Support a new parameter $intent_id, add one more test.
- Update to re-use the valid attached intent_id for UPE_Payment_Gateway

htdat committed:
For JS (UPE flow) Handle the redirection

- Update from handlePreviousOrderPaid to handleDuplicatePayments
- Add redirection for wcpay_upe_previous_successful_intent

htdat committed:
Add info text to the thank-you page

htdat committed:
Revert "Try to reuse the existing intent id."

This reverts commit 2df0db4868221ca2e7755cad4d4141b77282d188.

htdat committed:
Merge branch 'develop' into add/check-intent-attached-to-order-before-payment

htdat committed:
Add tests for WC_Payment_Gateway_WCPay_Process_Payment_Test

htdat committed:
Add tests for UPE_Payment_Gateway_Test

htdat committed:
Add changelog

htdat committed:
Merge branch 'develop' into add/check-intent-attached-to-order-before-payment

RadoslavGeorgiev commented on changelog/add-check-intent-attached-to-order-before-payment:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: update
+
+Check the status of attached intent during the payment processing.
```
This patch is pretty significant, I think making it a bit longer and non-technical could be benefitial.

How about something in the spirit of "Check the status of previously initiated payments and mark orders as processing instead of initiating a new payment".

htdat replied:
Makese sense, updated [a57a775](https://github.com/Automattic/woocommerce-payments/pull/5346/commits/a57a775cde59fec83353fff3b59703009f8222d7)

htdat commented:
FYI. I am still looking into this [E2E test failure](https://github.com/Automattic/woocommerce-payments/actions/runs/3827260909/jobs/6511697984), which seems consistent: 

> FAIL tests/e2e/specs/subscriptions/shopper/shopper-subscriptions-manage-payments.spec.js (125.827s)
>   Shopper > Subscriptions > Manage Payment Methods
>     ✓ should be able to purchase a subscription (13312ms)
>     ✓ should change a default payment method to a new one (8541ms)
>     ✕ should set a payment method to an already saved card (100007ms)

## Update

This is [debug.log](https://github.com/Automattic/woocommerce-payments/files/10338350/debug.log) from the E2E run above. 

> [03-Jan-2023 06:10:49 UTC] PHP Fatal error:  Uncaught Exception: Error: No route was found matching the URL and request method. in /var/www/html/wp-content/plugins/woocommerce-payments/includes/class-wc-payment-gateway-wcpay.php:797
> Stack trace:
> #0 /var/www/html/wp-content/plugins/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-change-payment-gateway.php(334): WC_Payment_Gateway_WCPay->process_payment(93)

This error happens when a customer is trying to change a subscription (not order, its post type is `shop_subscription` in wp_posts) from the current payment method to another existing payment. 

With a subscription, we're saving a setup intent (starting with `seti_`) as order meta  `_intent_id`. Without checking this fact, we will send a GET request for a setup intention to the payment intention endpoint. 
https://github.com/Automattic/woocommerce-payments/blob/b5566a79245c48cb55324ec7f4279615f9ddd59e/includes/class-wc-payment-gateway-wcpay.php#L1951-L1951

I fixed it in this commit by checking intent ID prefix before processing further. https://github.com/Automattic/woocommerce-payments/pull/5346/commits/f3f7b5ef6f7655b82b980f8de70c5cd301258eeb

RadoslavGeorgiev commented on client/checkout/api/index.js:
```diff
@@ -695,14 +695,19 @@ export default class WCPayAPI {
+		const hasFlag =
+			response.wcpay_upe_paid_for_previous_order ||
+			response.wcpay_upe_previous_successful_intent;
+		if ( hasFlag && response.redirect ) {
```
I see how this works, and it makes sense, but I would maybe try making it a bit more readable:

```js
/**
 * Redirect to the order-received page for duplicate payments:
 *
 * @param {Object} response Response data to check if doing the redirect.
 * @return {boolean} Returns true if doing the redirection.
 */
handleDuplicatePayments( {
	wcpay_upe_paid_for_previous_order: previouslyPaid,
	wcpay_upe_previous_successful_intent: previousSuccessfulIntent,
	redirect,
} ) {
	if ( redirect ) {
		// Another order has the same cart content and was paid.
		if ( previouslyPaid ) {
			return ( window.location = redirect );
		}

		// Another intent has the equivalent successful status for the order.
		if ( previousSuccessfulIntent ) {
			return ( window.location = redirect );
		}
	}

	return false;
}
```

The change is mostly to bring comments closer to the code, I'm not insisting on it.

htdat replied:
Great suggestion! Updated https://github.com/Automattic/woocommerce-payments/pull/5346/commits/290c0d2660e552291d61cc7a5963192fb0208baa

htdat committed:
Specify that we will check only payment intent.

Fix this https://github.com/Automattic/woocommerce-payments/pull/5346#issuecomment-1369656021

htdat committed:
Update tests to group test cases better in the upcoming commits.

htdat committed:
Add tests for invalid PI ids

htdat committed:
Re-arrange test positions to make more sense

RadoslavGeorgiev commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1919,6 +1929,51 @@ protected function get_account_branding_icon( $default_value = '' ): string {
+		}
+
+		$intent        = $this->payments_api_client->get_intent( $intent_id );
+		$intent_status = $intent->get_status();
```
An exception might occur during the `$api_client->get_intent()` call or if the response was erroneous, while trying to get the intent status. Right now if either happened, the whole payment process would fail, even though the exception would be caught, and the handling will be graceful.

Nevertheless, I'm curious if it makes sense to handle the exception here by just logging it, and returning `null`.

htdat replied:
> Nevertheless, I'm curious if it makes sense to handle the exception here by just logging it, and returning null.

Agreed with this. Updated https://github.com/Automattic/woocommerce-payments/pull/5346/commits/cf3918ee6a76328db4f3322889c68357535c516f

RadoslavGeorgiev reviewed (changes requested):
Code goods very well overall, a couple of nitpicks, but the approach seems bulletproof 🥳 

Just curious, are you planning to move the saving/storage of intents closer to API calls in a different PR?

htdat committed:
Update changelog message.

htdat committed:
Update handleDuplicatePayments based on feedback

htdat committed:
Add try/catch when fetching intent

htdat committed:
Merge branch 'develop' into add/check-intent-attached-to-order-before-payment

htdat commented:
Thanks for the awesome feedback! 

@RadoslavGeorgiev - I've addressed your comments, and accepted all of them. This PR is ready for review again. 

Regarding: 

> Just curious, are you planning to move the saving/storage of intents closer to API calls in a different PR?

Good question but I would have mentioned earlier. In UPE gateway, `attach_intent_info_to_order` in [process_payment](https://github.com/Automattic/woocommerce-payments/blob/cf3918ee6a76328db4f3322889c68357535c516f/includes/payment-methods/class-upe-payment-gateway.php#L531) and [process_redirect_payment](https://github.com/Automattic/woocommerce-payments/blob/cf3918ee6a76328db4f3322889c68357535c516f/includes/payment-methods/class-upe-payment-gateway.php#L705) has already been close to the API calls 

For the main gateway, in this attach method is also already close to the API call in [update_order_status](https://github.com/Automattic/woocommerce-payments/blob/cf3918ee6a76328db4f3322889c68357535c516f/includes/class-wc-payment-gateway-wcpay.php#L2623-L2624) already. 

The only remaining attach method in [process_payment_for_order](https://github.com/Automattic/woocommerce-payments/blob/cf3918ee6a76328db4f3322889c68357535c516f/includes/class-wc-payment-gateway-wcpay.php#L1232) of the main gateway can be moved earlier, right after [create_and_confirm_intention](https://github.com/Automattic/woocommerce-payments/blob/cf3918ee6a76328db4f3322889c68357535c516f/includes/class-wc-payment-gateway-wcpay.php#L1065-L1090). However, reviewing all the complex code between the current position and the potential earlier position, I think it's acceptable to keep it in the current position. Otherwise, there are some pitfalls: 

- In many cases, we still need to call this attach method in the same position, then that's double database update calls.
- The code block between this method and two other methods `update_order_status_from_intent` and `attach_exchange_info_to_order` will be separated while they would be close to each other.

htdat commented:
> E2E Tests - Pull Request / WC - latest | blocks - shopper (pull_request) - [Details](https://github.com/Automattic/woocommerce-payments/actions/runs/3837591347/jobs/6533937785)

This E2E test failure is similar to a recent scheduled E2E run for the `develop` branch - [E2E Tests - All](https://github.com/Automattic/woocommerce-payments/actions/runs/3837809250).

htdat commented:
> E2E Tests - Pull Request / WC - latest | blocks - shopper (pull_request) - [Details](https://github.com/Automattic/woocommerce-payments/actions/runs/3837591347/jobs/6533937785)

This E2E test failure is similar to a recent scheduled E2E run for the `develop` branch - [E2E Tests - All](https://github.com/Automattic/woocommerce-payments/actions/runs/3837809250).

_Update_: Discussion regarding this: p1672840161807159-slack-C04D1QU76PK

htdat committed:
Merge branch 'develop' into add/check-intent-attached-to-order-before-payment

RadoslavGeorgiev reviewed (approved):
The code looks and works great 🥳 🚢 

I've tested:
- UPE & non-UPE
- Blocks vs classic
- 3DS vs non-3DS. When testing with a 3DS card, I let it fail and then switched to 4242 to make sure the original payment method remains.

> In many cases, we still need to call this attach method in the same position, then that's double database update calls.

I think it's still worth simply storing the intent ID as early as possible. `$order->update_meta_data()` does not trigger a database update call by itself, that only happens when `$order->save()` or `$order->save_meta_data()` is called. Those would be called before marking the order as failed anyway, so adding a single update method call after the API call would make a lot of sense without any practical overhead. I'll leave that up to you though, the PR is approved anyway :)

htdat committed:
Merge branch 'develop' into add/check-intent-attached-to-order-before-payment
